storage/dataflux/next_page.go

```diff
@@ -0,0 +1,184 @@
+	// endRange is the end offset of the objects to be listed.
+	endRange string
+	// bucketHandle is the bucket handle of the bucket from which objects are to be listed.
+	bucketHandle *storage.BucketHandle
+	// query is the storage.Query to filter objects for listing.
+	query storage.Query
+	// skipDirectoryObjects is to indicate whether to skip or list directory objects.
+	skipDirectoryObjects bool
+	// generation is the generation number of the last object in the page.
+	generation int64
+}
+
+// nextPageResult represents the results of fetching a single page of objects
+// from a GCS listing operation and information for remaining objects to be listed.
+type nextPageResult struct {
+	// items is the list of objects listed.
+	items []*storage.ObjectAttrs
+	// doneListing indicates whether the lister has completed listing.
+	doneListing bool
+	// nextStartRange is the start offset of the next page of objects to be listed.
+	nextStartRange string
+	// generation is the generation number of the last object in the page.
+	generation int64
+}
+
+// nextPage retrieves a single page of objects from GCS using the provided
+// listing options (nextPageOpts). It returns a nextPageResult containing the
+// list of objects, a flag indicating if the listing is complete, the starting
+// point for the next page, and the generation of the last object in the page.
+// In case multiple versions of objects needs to be listed, then
+func nextPage(ctx context.Context, opts nextPageOpts) (*nextPageResult, error) {
+
+	opts.query.StartOffset = addPrefix(opts.startRange, opts.query.Prefix)
+	opts.query.EndOffset = addPrefix(opts.endRange, opts.query.Prefix)
+	objectIterator := opts.bucketHandle.Objects(ctx, &opts.query)
+	var items []*storage.ObjectAttrs
+
+	// nameLexLast is the name of lexicographically last object in the page.
+	nameLexLast := ""
+	// indexLexLast is the index of lexicographically last object in the page.
+	// If the item is iterated but not added to the items list, then indexLexLast is -1.
+	indexLexLast := 0
+	// indexItemLast is the index of the last item in the items list.
+	indexItemLast := -1
+
+	// The Go Listing API does not expose an interface to list multiple objects together,
+	// thus we need to manually loop to construct a page of results using the iterator.
+	for i := 0; i < wsDefaultPageSize; i++ {
+		attrs, err := objectIterator.Next()
+
+		// If the lister has listed the last item for the assigned range,
+		// then set doneListing to true and return.
+		if err == iterator.Done {
+			return &nextPageResult{
+				items:          items,
+				doneListing:    true,
+				nextStartRange: "",
+				generation:     int64(0),
+			}, nil
+		} else if err != nil {
+			return nil, fmt.Errorf("iterating through objects: %w", err)
+		}
+
+		// Skip object versions already processed in the previous page to prevent duplicates.
+		if opts.query.Versions && opts.query.StartOffset == attrs.Name && attrs.Generation < opts.generation {
```

Comment:
Let's make sure we test versions against prod.  

Why are we checking if `attrs.Generation < opts.generation`?  I don't think we can assume objects listed will be listed in increasing generation order.

Reply:
In versions parameter it shows objects will be listed in order of increasing generation number.

https://cloud.google.com/storage/docs/json_api/v1/objects/list#parameters

Reply:
Ah, my bad, TIL. In that case this looks good 👍